File: app/src/main/java/com/wojdor/hearthstonecards/application/base/BaseAndroidViewModel.java
package com.wojdor.hearthstonecards.application.base;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.support.annotation.NonNull;

import com.wojdor.hearthstonecards.data.database.CardDatabase;
import com.wojdor.hearthstonecards.data.database.dao.CardDao;
import com.wojdor.hearthstonecards.data.service.CardApi;
import com.wojdor.hearthstonecards.data.service.CardService;
import com.wojdor.hearthstonecards.data.session.UserSession;

import io.reactivex.disposables.CompositeDisposable;

public abstract class BaseAndroidViewModel extends AndroidViewModel {

    protected CardApi cardApi;
    protected CompositeDisposable disposable;
    protected CardDao cardDao;
    protected UserSession userSession;

    public BaseAndroidViewModel(@NonNull Application application) {
        super(application);
        cardApi = CardService.getInstance();
        disposable = new CompositeDisposable();
        cardDao = CardDatabase.getInstance(application).cardDao();
        userSession = UserSession.getInstance(application);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposable.clear();
    }
}

File: app/src/main/java/com/wojdor/hearthstonecards/application/base/BaseAndroidViewModel.kt
package com.wojdor.hearthstonecards.application.base

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import com.wojdor.hearthstonecards.application.util.CardImageDownloader
import com.wojdor.hearthstonecards.data.database.CardDatabase
import com.wojdor.hearthstonecards.data.repository.CardRepository
import com.wojdor.hearthstonecards.data.service.CardService
import com.wojdor.hearthstonecards.data.session.UserSession

abstract class BaseAndroidViewModel(application: Application) : AndroidViewModel(application) {

    protected val repository by lazy {
        // TODO: Use koin to inject everything!
        CardRepository.getInstance(CardService.getInstance(), CardDatabase.getInstance(application).cardDao(),
                UserSession.getInstance(application), CardImageDownloader(application))
    }

    override fun onCleared() {
        super.onCleared()
        repository.onDestroy()
    }
}
